pkg/dogstatsd/server.go

```diff
@@ -466,6 +490,80 @@ func (s *Server) storeMetricStats(sample metrics.MetricSample) {
 	ms.Name = sample.Name
 	ms.Tags = strings.Join(sample.Tags, " ") // we don't want/need to share the underlying array
 	s.Debug.Stats[key] = ms
+
+	s.Debug.metricsCounts.metricChan <- struct{}{}
+}
+
+// EnableMetricsStats enables the debug mode of the DogStatsD server and start
+// the debug mainloop collecting the amount of metrics received.
+func (s *Server) EnableMetricsStats() {
+	s.Debug.Lock()
+	defer s.Debug.Unlock()
+
+	// already enabled?
+	if atomic.LoadUint64(&s.Debug.Enabled) == 1 {
+		return
+	}
+
+	atomic.StoreUint64(&s.Debug.Enabled, 1)
+	go func() {
+		ticker := time.NewTicker(time.Millisecond * 100)
+		var closed bool
+		log.Info("Starting the DogStatsD debug loop.") // TODO(remy): move this to Debug
+		for {
+			mc := s.Debug.metricsCounts
+			select {
+			case <-ticker.C:
+				sec := time.Now().Truncate(time.Second)
+				if sec.After(mc.currentSec) {
+					mc.currentSec = sec
+					mc.bucketIdx++
+					if mc.bucketIdx >= len(mc.counts) {
+						mc.bucketIdx = 0
+					}
+					// compare this one to the sum of all others
+					// if the difference is higher than all others sum, consider this
+					// as an anomaly.
+					var sum int64
+					for _, v := range mc.counts {
+						sum += v
+					}
+					sum -= mc.counts[mc.bucketIdx]
+					if mc.counts[mc.bucketIdx] > sum {
+						log.Warnf("A spike, a large increase or a large decrease of metrics has been observed by DogStatSd: here is the last 5 seconds count of metrics: %v", mc.counts)
+					}
```

Comment:
I probably miss something but I cannot see how does it work when the number of metric increase.

Consider the following case where the first 2 seconds there is 1 metric by second and then 100 metrics by second.

|`ticker.C` call count. | count | sum line 532 | mc.counts[mc.bucketIdx] line 532 | mc.bucketIdx after line 520|
| ---------------- | ----- | ------------ | ---------------------------| -------- |
| 1  | [1,0,0,0,0] | 1 | 0 | 1 |
| 2  | [1,1,0,0,0] | 2 | 0 | 2 |
| 3  | [1,1,100,0,0] | 102 | 0 | 3 |
| 4  | [1,1,100,100,0] | 202 | 0 | 4 |
| 5  | [1,1,100,100,100] | 301 | 1 | 0 |
| 6  | [100,1,100,100,100] | 400 | 1 | 1 |
| 7  | [100,100,100,100,100] | 400 | 100 | 2 |

Where is my mistake?

Reply:
You're right, bucketIdx was incremented too soon, fixed 👍